### ISSUE TITLE
Pascal, Hilbert and/or other special matrices in `jax.scipy.linalg`

### ISSUE THREAD

**Issue by pnkraemer** [google/jax#10144]
Hi! 

`scipy.linalg` implements ![special matrices](https://docs.scipy.org/doc/scipy/reference/linalg.html#special-matrices). At the moment, `jax.scipy.linalg` does not do that, but it would be useful for me to have access to Pascal and Hilbert matrices via `jax`. (I need them for some state-space modelling stuff, if that is relevant information.)

I could make a pull request if it helps :)
As a starting point, I will explain what I am doing currently below, and I will also explain some difficulties with `jit`ting that might need to be resolved before a pull request? 

## Related issues 

* https://github.com/google/jax/issues/1646 (vaguely related I guess)
* https://github.com/google/jax/issues/3171 (see below for why this is related)

## Hilbert matrix

For example, a Hilbert matrix would be rather straightforward to implement, via 
```python
def hilbert(n):
    a = jnp.arange(n)
    return 1 / (a[:, None] + a[None, :] + 1)
```
A problem I could see with this implementation is appropriate `jit` and `vmap` behaviour: the size of the output depends on `n` :( . This is true for most special matrices, unfortunately.

Therefore, in my own code, I have split the function into something like
```python

def hilbert(n):
    a = jnp.arange(n)
    return _hilbert(a)

@jax.jit
def _hilbert(a):
    return 1 / (a[:, None] + a[None, :] + 1)
```
which at least allows to `jit` parts of the implementation. 
An alternative would be to provide a Hankel matrix (where the shape of the matrix is determined by the shapes of the input vectors, and which should be easier to `jit`), and then let users like me do the Hilbert-via-Hankel conversion depending on each application. Scipy's Hankel matrix, however, uses `np.lib.as_strided`, and I understand from issue #3171 that this is not intended to be supported.
But this can probably be resolved, if one decides to go down that road.

## Pascal matrix

Pascal matrices are a little less obvious to implement than Hankel/Hilbert matrices, because they use binomial coefficients. For my applications, I have so far computed those via `jax.lax.exp(jax.lax.lgamma(n+1.))` (because `jnp.prod(jnp.arange(1, n+1))` has `jit` issues again; if I remember correctly, I took the gamma-function-idea from `jet.py`) and then vmapped a factorial implementation via 
```python
def pascal(n):
    return _pascal(jnp.arange(n))

@jax.jit
def _pascal(a):
    return _binom(a[:, None], a[None, :])


def _broadcast_to_matrix(k):
    """Take a function f(scalar, scalar)-> scalar and return a function g((n,m) array, (m, k) array) -> (n,k) array"""
    k_vmapped_x = jax.vmap(k, in_axes=(0, None), out_axes=-1)
    k_vmapped_xy = jax.vmap(k_vmapped_x, in_axes=(None, 1), out_axes=-1)
    return jax.jit(k_vmapped_xy)


@_broadcast_to_matrix
@jax.jit
def _binom(n, k):
    a = _factorial(n)
    b = _factorial(n - k)
    c = _factorial(k)
    return a / (b * c)


@jax.jit
def _factorial(n):
    return jax.lax.exp(jax.lax.lgamma(n + 1.0))
```
This is slightly different to `scipy.linalg.pascal`, however, because my implementation returns only floats and a general Pascal matrix should do integers, at least for `scipy.linalg.pascal(..., exact=True)`. And the implementation would be fairly different to ![how scipy does it](https://github.com/scipy/scipy/blob/v1.8.0/scipy/linalg/_special_matrices.py#L776-L859), which might also not be desired.

## Long story short
Hilbert and Pascal matrices would be useful to have, and if the jitting issues can be resolved (or do not matter too much?) I can make a pull request. 

What do you think?

**Comment by jakevdp** [comment]
I think these would be useful contributions. Regarding jit-compatibility, You might think about pre-compiling these with `static_argnames` (search through the JAX source to see other examples of this).

If you're willing to put together a PR, I'd be happy to review!

**Comment by zhangqiaorjc** [comment]
We seem to have a `static_argnames` example in `jax.jit` documentation itself, see the last example https://jax.readthedocs.io/en/latest/_autosummary/jax.jit.html?highlight=static_argnames#jax.jit

**Comment by pnkraemer** [comment]
Thanks for the suggestions! I will open a PR soon then, and let you know in case I have more questions.

**Comment by pnkraemer** [comment]
#10161 adds the Hilbert matrix. Once this is through, I will add the Pascal matrix as well (but since its implementation might be less obvious, I decided to split them into 2 PRs.)

**Comment by aaronatp** [comment]
Hi @jakevdp I hope you're doing well. Can I open a PR for this issue?

**Comment by haniyakonain** [comment]
Hi! 👋
As part of my Google Summer of Code 2025 (GSoC) proposal with JAX, I’ve been working on adding special matrices to jax.scipy.linalg. Here's an update along with the solution to an error I encountered during development.

✅ Hilbert Matrix
Implemented the Hilbert matrix using JAX-compatible operations:

```
import jax
import jax.numpy as jnp

@jax.jit
def _hilbert_impl(a):
    return 1 / (a[:, None] + a[None, :] + 1)

def hilbert(n: int) -> jnp.ndarray:
    a = jnp.arange(n)
    return _hilbert_impl(a)
```

Example Output:

```
>>>hilbert(3)
Array([[1.        , 0.5       , 0.33333334],
       [0.5       , 0.33333334, 0.25      ],
       [0.33333334, 0.25      , 0.2       ]], dtype=float32)
```

✅ Pascal Matrix
Implemented using jax.lax.lgamma for stable binomial coefficient calculation with jit compatibility:

```

@jax.jit
def _factorial(n):
    return jnp.exp(jax.lax.lgamma(n + 1.0))

@jax.jit
def _binom(n, k):
    return _factorial(n) / (_factorial(k) * _factorial(n - k))

def pascal(n: int) -> jnp.ndarray:
    i = jnp.arange(n)[:, None]
    j = jnp.arange(n)[None, :]
    return _binom(i + j, j)
```
Example Output:

```
>>> pascal(4)
Array([[ 1.,  1.,  1.,  1.],
       [ 1.,  2.,  3.,  4.],
       [ 1.,  3.,  6., 10.],
       [ 1.,  4., 10., 20.]], dtype=float32)
```

🛠️ Error Solved: pjit in_axis_resources must be a prefix of the argument pytree
While testing these implementations with batching, I encountered:
`
ValueError: pjit in_axis_resources must be a prefix of the argument pytree`

✅ Fix:
I updated the implementation to skip None entries when matching in_axis_resources with the actual input pytree. This ensures compatibility with pjit, jit, and batched inputs.

📌 Next Steps
 Add both functions to jax/scipy/linalg/ (e.g., special_matrices.py)

 Write tests under tests/scipy/linalg_tests/

 Expose the functions in the corresponding __init__.py

Would love your feedback before opening a PR! Thanks! 🙌

**Comment by Vedag812** [comment]
Hi @jakevdp! I'd like to work on adding more special matrices. 
I see Pascal matrix was recently added. Are there other special matrices 
(like Companion, Leslie, or Fiedler) that would be good to add?
I'm comfortable with linear algebra and have worked with NumPy/JAX.
Would love your guidance on which one to start with.

**Comment by Sagnik2003** [comment]
Hi @pnkraemer I see that Pascal, Hilbert matrices are already done, so I would like to work on other special matrices (such as DFT matrix, Hadamard matrix, inverse of Hadamard matrix etc.) since I already have a little theoretical ideas about some of them

**Comment by Harshadev-24** [comment]
Hi! I’d like to open a new PR adding another special matrix to `jax.scipy.linalg`.

This issue mentions Hankel as an alternative to Hilbert/Pascal, but SciPy’s implementation uses `np.lib.as_strided` (which JAX won’t support). I can implement `jax.scipy.linalg.hankel` using pure JAX indexing/broadcasting (no as_strided), matching SciPy’s API (first column `c`, last row `r`).

Is anyone currently working on Hankel (or similar matrices like circulant/leslie/invhilbert)? If not, I’ll open a draft PR in the next day or two.

**Comment by Akshat19G** [comment]
Hi, I'd like to work on this issue. Could I be assigned?

**Comment by jakevdp** [comment]
Hi @Akshat19G – we don't assign issues outside the core team. If you'd like to contribute, please feel free to open a pull request (and see our [contributing guidelines](https://docs.jax.dev/en/latest/contributing.html)).
